cpp/src/datetime/datetime_util.cpp

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2018-2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <cudf/cudf.h>
+#include <cudf/unary.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/datetime.hpp>
+
+namespace cudf {
+namespace datetime {
+
+std::pair<gdf_column, gdf_column> resolve_common_time_unit(gdf_column const& lhs,
+                                                           gdf_column const& rhs) {
+
+  gdf_column lhs_out = cudf::empty_like(lhs);
+  gdf_column rhs_out = cudf::empty_like(rhs);
+
+  if ((lhs.dtype == GDF_TIMESTAMP || lhs.dtype == GDF_DATE64) &&
+      (rhs.dtype == GDF_TIMESTAMP || rhs.dtype == GDF_DATE64)) {
+
+    gdf_time_unit lhs_unit = lhs.dtype_info.time_unit;
+    gdf_time_unit rhs_unit = rhs.dtype_info.time_unit;
+
+    // GDF_DATE64 and TIME_UNIT_NONE are always int64_t milliseconds
+    if (lhs.dtype == GDF_DATE64 || lhs_unit == TIME_UNIT_NONE) {
+      lhs_unit = TIME_UNIT_ms;
+    }
+    if (rhs.dtype == GDF_DATE64 || rhs_unit == TIME_UNIT_NONE) {
+      rhs_unit = TIME_UNIT_ms;
+    }
+    if (lhs_unit != rhs_unit) {
+      gdf_dtype_extra_info out_info{std::max(lhs_unit, rhs_unit)};
```

Comment:
I don't love that this is relying implicitly on the ordering of the enum values for the time unit resolutions. The order of those enums could always change, and then this code would break.

Reply:
I agree, but wasn't sure how stable to consider the types. I've exploded this out into a conditional block in the latest commit.